pin/gc/gc.go

```diff
@@ -33,146 +27,232 @@ type Result struct {
+	if criticalError != nil {
+		return
+	}
+
+	// Add white objects
+	keychan, err := g.bs.AllKeysChan(ctx)
 	if err != nil {
-		errors = true
 		output <- Result{Error: err}
+		return
 	}
 
-	bestEffortGetLinks := func(ctx context.Context, cid *cid.Cid) ([]*node.Link, error) {
-		links, err := ls.GetLinks(ctx, cid)
-		if err != nil && err != dag.ErrNotFound {
-			errors = true
-			output <- Result{Error: &CannotFetchLinksError{cid, err}}
+loop:
+	for {
+		select {
+		case c, ok := <-keychan:
+			if !ok {
+				break loop
+			}
+			tri.InsertFresh(c)
+		case <-ctx.Done():
+			fmt.Printf("ctx done\n")
+			output <- Result{Error: ctx.Err()}
+			return
 		}
-		return links, nil
 	}
-	err = Descendants(ctx, bestEffortGetLinks, gcs, bestEffortRoots)
-	if err != nil {
-		errors = true
-		output <- Result{Error: err}
+
+	// Take the lock
+	unlocker, elock := getGCLock(ctx, g.bs)
+
+	defer unlocker.Unlock()
+	defer elock.Done()
+
+	// Add the roots again, they might have changed
+	for _, r := range g.roots {
+		cids, err := r.Get()
+		if err != nil {
+			criticalError = err
+			return
+		}
+		for _, c := range cids {
+			if !r.Direct {
+				tri.InsertGray(c, r.Strict)
+			} else {
+				// this special case prevents incremental and concurrent GC
+				tri.blacken(c, enumStrict)
+			}
+		}
 	}
 
-	for _, k := range pn.DirectKeys() {
-		gcs.Add(k)
+	// Reenumerate, fast as most will be duplicate
+	for {
+		finished, err := tri.EnumerateStep(ctx, getLinks, bestEffortGetLinks)
+		if err != nil {
+			output <- Result{Error: err}
+			criticalError = ErrCannotFetchAllLinks
+		}
+		if finished {
+			break
+		}
 	}
 
-	err = Descendants(ctx, getLinks, gcs, pn.InternalPins())
-	if err != nil {
-		errors = true
-		output <- Result{Error: err}
+	if criticalError != nil {
+		return
+	}
+
+	emark.Done()
+	esweep := log.EventBegin(ctx, "GC.sweep")
+
+	var whiteSetSize, blackSetSize uint64
+
+loop2:
+	for v, e := range tri.colmap {
+		if e.getColor() != tri.white {
+			blackSetSize++
+			continue
+		}
+		whiteSetSize++
+
+		c, err := cid.Cast([]byte(v))
+		if err != nil {
+			// this should not happen
+			panic("error in cast of cid: " + err.Error())
+		}
+
+		err = g.bs.DeleteBlock(c)
+		if err != nil {
+			output <- Result{Error: &CannotDeleteBlockError{c, err}}
+			criticalError = ErrCannotDeleteSomeBlocks
+			continue
+		}
+		select {
+		case output <- Result{KeyRemoved: c}:
+		case <-ctx.Done():
+			break loop2
+		}
+	}
+
+	esweep.Append(logging.LoggableMap{
+		"whiteSetSize": fmt.Sprintf("%d", whiteSetSize),
+		"blackSetSize": fmt.Sprintf("%d", blackSetSize),
+	})
+	esweep.Done()
+}
+
+// Result represents an incremental output from a garbage collection
+// run.  It contains either an error, or the cid of a removed object.
+type Result struct {
+	KeyRemoved *cid.Cid
+	Error      error
+}
+
+func getGCLock(ctx context.Context, bs bstore.GCBlockstore) (bstore.Unlocker, *logging.EventInProgress) {
+	elock := log.EventBegin(ctx, "GC.lockWait")
+	unlocker := bs.GCLock()
```

Comment:
Yes, to prevent for example blocks for being added. In future we can have incremental GC where those pauses (this is what taking major lock by GC is called) are short but spread over time.

Reply:
Or to highlight the problem more cleanly, to prevent `ipfs add` from adding blocks that are referencing some white marked nodes.

Reply:
I see. I was thinking that that case might be equivalent to doing the GC before adding the block but that's not true because the fact that we had both blocks at the same time can be observed. Having a GCed system where something can go from unreachable (by a root) to reachable in the middle of a GC run is weird.